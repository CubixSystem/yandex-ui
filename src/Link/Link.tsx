import {
    AnchorHTMLAttributes,
    createElement,
    DetailedHTMLFactory,
    FC,
    MouseEventHandler,
    ReactNode,
    ReactType,
    RefObject,
} from 'react';
import { cn } from '@bem-react/classname';

import './Link.css';

export type ContainerElement = HTMLSpanElement | HTMLAnchorElement;

type ReactLinkElement = DetailedHTMLFactory<AnchorHTMLAttributes<ContainerElement>, ContainerElement>;

export interface ILinkProps {
    /**
     * Адрес ссылки. Если указано, то компонент будет оформлен тегом `a`, в противном случае — `span`.
     *
     * Значение игнорируется при использовании модификатора `pseudo`
     */
    href?: string;

    /**
     * Выключение интерактивности ссылки.
     * Состояние, при котором ссылка отображается, но недоступна для действий пользователя
     */
    disabled?: boolean;

    /**
     * Ссылка на корневой DOM-элемент компонента
     */
    innerRef?: RefObject<ContainerElement>;

    /**
     * Ссылка на DOM-элемент нативного контрола
     */
    controlRef?: RefObject<ContainerElement>;

    /**
     * Указание для отрисовки компонента
     */
    as?: ReactType;

    /**
     * Дополнительный класс
     */
    className?: string;

    /**
     * Содержимое ссылки
     */
    children?: ReactNode;

    /**
     * HTML-атрибут `title`
     */
    title?: string;

    /**
     * HTML-атрибут `target`
     */
    target?: string;

    /**
     * HTML-атрибут `rel`
     */
    rel?: string;

    /**
     * HTML-атрибут `tabIndex`. Определяет последовательность перехода между ссылками при нажатии на кнопку Tab
     */
    tabIndex?: number;

    /**
     * Обработчик события клика.
     */
    onClick?: MouseEventHandler<ContainerElement>;

    /**
     * HTML-атрибут `role`
     */
    role?: string;
}

export const cnLink = cn('Link');

/**
 * Компонент для создания ссылок.
 * @param {ILinkProps} props
 */
export const Link: FC<ILinkProps> = ({
    as: AsComponent,
    children,
    controlRef,
    disabled,
    innerRef,
    role,
    // Извлекаем свойства, т.к. они не нужны на DOM узле
    // FIXME: https://github.com/bem/bem-react/issues/381
    // @ts-ignore
    pseudo: _pseudo,
    // @ts-ignore
    view: _view,
    ...props
}) => {
    const className = cnLink(null, [props.className]);
    const Component = AsComponent || ((props.href ? 'a' : ('span' as any)) as ReactLinkElement);
    let relationship = props.rel;

    if (props.target === '_blank' && relationship !== undefined && relationship.indexOf('noopener') === -1) {
        // Пользовательский атрибут имеет больший приоритет.
        relationship = `${relationship} noopener`;
    }

    // PERF: Не используем TSX, чтобы избежать лишнего клонирования props.
    return createElement(
        Component,
        Object.assign(props, {
            'aria-disabled': disabled,
            className,
            ref: innerRef || controlRef,
            rel: relationship,
            role,
            tabIndex: disabled ? -1 : props.tabIndex,
        }),
        children,
    );
};

Link.displayName = cnLink();
