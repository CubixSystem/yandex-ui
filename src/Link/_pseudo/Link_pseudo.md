### pseudo

Делает ссылку в псевдоссылкой. Клик по такой ссылке не вызывает перехода на страницу, указанную в свойстве `href`, а вызывает событие, которое указано в свойстве `onClick`.

Псевдоссылки используются для того, чтобы открыть/скрыть какую-то информацию на странице, например показать всплывающую подсказку.

Для оформления псевдоссылок можно использовать тему `pseudo`.

<!-- props:start -->
| Свойство   | Тип                                                | Описание                       |
| ---------- | -------------------------------------------------- | ------------------------------ |
| pseudo?    | `false \| true`                                    | Псевдоссылка                   |
| onKeyDown? | `(event: KeyboardEvent<ContainerElement>) => void` | Обработчик события `onKeyDown` |
<!-- props:end -->
