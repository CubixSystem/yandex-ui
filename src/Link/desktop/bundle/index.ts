export * from '../../Link.bundle/desktop';
