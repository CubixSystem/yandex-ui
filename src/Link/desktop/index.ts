export * from '../Link@desktop';
export * from '../_pseudo/Link_pseudo';
export * from '../_theme/Link_theme_black@desktop';
export * from '../_theme/Link_theme_ghost@desktop';
export * from '../_theme/Link_theme_normal@desktop';
export * from '../_theme/Link_theme_outer@desktop';
export * from '../_theme/Link_theme_pseudo@desktop';
export * from '../_theme/Link_theme_strong@desktop';
export * from '../_view/Link_view_default@desktop';
