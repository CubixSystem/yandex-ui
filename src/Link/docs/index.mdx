---
id: Link
path: /components/link/examples

title: Link
description: Компонент для создания ссылок.

links:
 - label: source
   url: https://github.yandex-team.ru/search-interfaces/frontend/tree/master/packages/lego-components/src/Link
 - label: startrek
   url: https://search.yandex-team.ru/stsearch?text=Link.ts&facet.queue=ISL&facet.type=bug&facet.status=128
 - label: figma
   url: https://www.figma.com/file/CaoHFEO0w6PaE8cVaYFDoq8i/L-E-G-O?node-id=1473%3A9880
 - label: storybook
   url: https://lego.yandex-team.ru/components/lego/latest/?storybook=%2Fstory%2Fnavigation-link-desktop--playground

tabs:
 - Examples
 - Code
---
