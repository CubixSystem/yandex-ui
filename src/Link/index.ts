export * from './Link';
export * from './_pseudo/Link_pseudo';
export * from './_theme/Link_theme_black';
export * from './_theme/Link_theme_ghost';
export * from './_theme/Link_theme_normal';
export * from './_theme/Link_theme_outer';
export * from './_theme/Link_theme_pseudo';
export * from './_theme/Link_theme_strong';
export * from './_view/Link_view_default';
