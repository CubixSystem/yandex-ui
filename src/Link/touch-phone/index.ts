export * from '../Link';
export * from '../_pseudo/Link_pseudo';
export * from '../_theme/Link_theme_black';
export * from '../_theme/Link_theme_ghost';
export * from '../_theme/Link_theme_normal@touch-phone';
export * from '../_theme/Link_theme_outer';
export * from '../_theme/Link_theme_pseudo@touch-phone';
export * from '../_theme/Link_theme_strong@touch-phone';
export * from '../_view/Link_view_default';
