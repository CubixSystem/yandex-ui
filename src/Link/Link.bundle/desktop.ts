import { compose, composeU } from '@bem-react/core';

// base
import { Link as LinkDesktop } from '../Link@desktop';
// _pseudo
import { withPseudo } from '../_pseudo/Link_pseudo';
// _theme
import { withThemeBlack } from '../_theme/Link_theme_black@desktop';
import { withThemeGhost } from '../_theme/Link_theme_ghost@desktop';
import { withThemeNormal } from '../_theme/Link_theme_normal@desktop';
import { withThemeOuter } from '../_theme/Link_theme_outer@desktop';
import { withThemePseudo } from '../_theme/Link_theme_pseudo@desktop';
import { withThemeStrong } from '../_theme/Link_theme_strong@desktop';
// _view
import { withViewDefault } from '../_view/Link_view_default@desktop';

export * from '../Link@desktop';

export const Link = compose(
    withPseudo,
    withViewDefault,
    composeU(withThemeBlack, withThemeGhost, withThemeNormal, withThemeOuter, withThemePseudo, withThemeStrong),
)(LinkDesktop);
