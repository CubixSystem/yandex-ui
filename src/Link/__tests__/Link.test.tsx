import React, { createRef } from 'react';

import { createClientRender, screen, fireEvent } from '../../internal/testing';
import { Link as LinkCommon } from '../Link';
import { Link as LinkDesktop } from '../desktop';
import { Link as LinkTouchPad } from '../touch-pad';
import { Link as LinkTouchPhone } from '../touch-phone';
import { withPseudo } from '../_pseudo/Link_pseudo';

const platforms = [
    ['desktop', withPseudo(LinkDesktop)],
    ['touch-pad', withPseudo(LinkTouchPad)],
    ['touch-phone', withPseudo(LinkTouchPhone)],
] as const;

describe.each(platforms)('Link@%s', (_platform, Link) => {
    const render = createClientRender();

    test('должен вернуть полный шаблон компонента (snapshot)', () => {
        const { container } = render(
            <Link href="www.test.ru" className="mix">
                Content
            </Link>,
        );

        expect(container.firstChild).toMatchSnapshot();
    });

    test('должен рендерить псевдо ссылку', () => {
        const { container } = render(
            <Link href="www.test.ru" pseudo>
                Content
            </Link>,
        );

        expect(container.firstChild).toMatchSnapshot();
    });

    test('должен быть установлен корректный displayName', () => {
        expect(LinkCommon.displayName).toBe('Link');
    });

    test('должен устанавливать ссылку на корневой DOM элемент', () => {
        const innerRef = createRef<HTMLElement>();
        render(<Link innerRef={innerRef} />);

        expect(innerRef.current).not.toBe(null);
    });

    // TODO: При чем здесь инпут непонятно, но ссылка реально поддерживает controlRef
    test('должен устанавливать ссылку на нативный DOM инпут', () => {
        const controlRef = createRef<HTMLInputElement>();
        render(<Link controlRef={controlRef} />);

        expect(controlRef.current).not.toBe(null);
    });

    test('должен вызывать событие onClick с event при клике', () => {
        const onClick = jest.fn();
        render(<Link role="link" onClick={onClick} />);

        fireEvent.click(screen.getByRole('link'));

        expect(onClick).toHaveBeenCalledTimes(1);
    });

    test('должен рендерить Link с вложенным content', () => {
        render(
            <Link>
                <i data-testid="link-content" />
            </Link>,
        );

        expect(screen.getByTestId('link-content')).toBeInTheDocument();
    });

    test('должен установить tabIndex', () => {
        const { setProps } = render(<Link role="link" tabIndex={1} />);

        expect(screen.getByRole('link')).toHaveAttribute('tabindex', '1');

        setProps({ disabled: true });

        expect(screen.getByRole('link')).toHaveAttribute('tabindex', '-1');
    });
});
