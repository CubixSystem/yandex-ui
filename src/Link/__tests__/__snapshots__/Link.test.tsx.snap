// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Link@desktop должен вернуть полный шаблон компонента (snapshot) 1`] = `
<a
  class="Link mix"
  href="www.test.ru"
>
  Content
</a>
`;

exports[`Link@desktop должен рендерить псевдо ссылку 1`] = `
<a
  class="Link Link_pseudo"
  href="www.test.ru"
>
  Content
</a>
`;

exports[`Link@touch-pad должен вернуть полный шаблон компонента (snapshot) 1`] = `
<a
  class="Link mix"
  href="www.test.ru"
>
  Content
</a>
`;

exports[`Link@touch-pad должен рендерить псевдо ссылку 1`] = `
<a
  class="Link Link_pseudo"
  href="www.test.ru"
>
  Content
</a>
`;

exports[`Link@touch-phone должен вернуть полный шаблон компонента (snapshot) 1`] = `
<a
  class="Link mix"
  href="www.test.ru"
>
  Content
</a>
`;

exports[`Link@touch-phone должен рендерить псевдо ссылку 1`] = `
<a
  class="Link Link_pseudo"
  href="www.test.ru"
>
  Content
</a>
`;
